spyder/plugins/ipythonconsole/widgets/kernel_connection_manager.py

```diff
@@ -0,0 +1,409 @@
+        self.shutdown_thread = None
+
+    def _thread_shutdown_kernel(self):
+        """Shutdown kernel."""
+        with self._shutdown_lock:
+            # Avoid calling shutdown_kernel on the same manager twice
+            # from different threads to avoid crash.
+            if self.kernel_manager.shutting_down:
+                return
+            self.kernel_manager.shutting_down = True
+        try:
+            self.kernel_manager.shutdown_kernel()
+        except Exception:
+            # kernel was externally killed
+            pass
+
+    def wait_shutdown_thread(self):
+        """Wait shutdown thread."""
+        thread = self.shutdown_thread
+        if thread is None:
+            return
+        if thread.isRunning():
+            try:
+                thread.kernel_manager._kill_kernel()
+            except Exception:
+                pass
+            thread.quit()
+            thread.wait()
+
+    def copy(self):
+        """Copy kernel."""
+        # Copy kernel infos
+        new_kernel = self.__class__(
+            connection_file=self.connection_file,
+            kernel_manager=self.kernel_manager,
+            known_spyder_kernel=self.known_spyder_kernel,
+            hostname=self.hostname,
+            sshkey=self.sshkey,
+            password=self.password,
+        )
+
+        # Copy std file
+        if self.stderr_obj is not None:
+            new_kernel.stderr_obj = self.stderr_obj.copy()
+        if self.stdout_obj is not None:
+            new_kernel.stdout_obj = self.stdout_obj.copy()
+        if self.fault_obj is not None:
+            new_kernel.fault_obj = self.fault_obj.copy()
+
+        # Get new kernel_client
+        new_kernel.init_kernel_client()
+        return new_kernel
+
+    def remove_files(self):
+        """Remove std files."""
+        for obj in [self.stderr_obj, self.stderr_obj, self.fault_obj]:
+            if obj is not None:
+                obj.remove()
+
+    def open_comm(self, kernel_comm):
+        """Open kernel comm"""
+        kernel_comm.open_comm(self.kernel_client)
+        self.kernel_comm = kernel_comm
+
+    def replace_std_files(self):
+        """Replace std files."""
+        for obj in [self.stderr_obj, self.stderr_obj, self.fault_obj]:
+            if obj is None:
+                continue
+            obj.remove()
+            fn = obj.filename
+            m = re.match(r"(.+_)(\d+)(.[a-z]+)", fn)
+            if m:
+                # Already a replaced file
+                path, n, ext = m.groups()
+                obj.filename = path + str(1 + int(n)) + ext
+                continue
+            m = re.match(r"(.+)(.[a-z]+)", fn)
+            if m:
+                # First replaced file
+                path, ext = m.groups()
+                obj.filename = path + "_1" + ext
+                continue
+            # No extension, should not happen
+            obj.filename += "_1"
+
+
+class CachedKernelMixin:
```

Comment:
Do you plan to use this mixin in other classes? I'm asking because I can see that it's only used in `IPythonConsoleWidget`, but the purpose of mixins is inheriting them in several classes.